articles/unique_network_flows.md

```diff
@@ -0,0 +1,171 @@
+
+Each publisher or subscriber that requires a specific QoS from the network is required to set the `flow_id` parameter to value unique within the node. The `flow_id` can be computed internally in the node using simple UID generation schemes (atomic integer counters or random integer sampling without duplicates) or supplied by an external component.
+
+The example C++ snippet below shows a node creating a publisher (`pub1`)  and a subscriber (`sub1`) with unique `flow_id` values. Subscriber `sub2` is created with the default Network Flow QoS policy (`flow_id` = 0).
+
+```cpp
+auto pub1_qos = rclcpp::QoS(rclcpp::KeepLast(10)).network_flow(0xf4688c52);
+pub1 = create_publisher<std_msgs::msg::UInt64>("pub1_topic", pub1_qos);
+
+auto sub1_qos = rclcpp::QoS(rclcpp::KeepLast(1)).network_flow(0x88181c45);
+sub1 = create_subscription<std_msgs::msg::String>("sub1_topic", sub1_qos, std::bind(&new_sub1_message, this, _1));
+
+auto sub2_qos = rclcpp::QoS(rclcpp::KeepLast(1));
+sub2 = create_subscription<std_msgs::msg::String>("sub2_topic", sub2_qos, std::bind(&new_sub2_message, this, _1));
+```
+
+The RMW implementation has several options to convert `flow_id` to a unique flow identifier visible in packet headers. We list three candidate options.
+
+If the node is communicating using IPv6, then the lower 20-bits of `flow_id` can be transferred to the Flow Label field. This creates a unique 3-tuple.
+
+Else, if the node is communicating via IPv4, then there are two alternatives. One is copy the `flow_id` to the Options field. Another is to copy the lower 6-bits of the `flow_id` to the DSCP field. Both alternatives enable the network to infer a custom 6-tuple (traditional 5-tuple plus Options/DSCP) that uniquely identifies flows.
+
+Both DDS and non-DDS RMW implementations can trivially set fields in IP headers using native socket API on all ROS2 platforms (Linux, Windows, MacOS). 
+
+### Option 2: Publisher/subscriber delegates flow ID generation to RMW
+
+This is a programmer-friendly alternative that modifies the Network Flow QoS policy to accept a boolean parameter called `unique_flow`. Setting `unique_flow` to `true` instructs the RMW implementation to generate a unique flow identifier for the publisher/subscriber associated with the Network Flow QoS policy. By default, `unique_flow` is set to `false`.
+
+The RMW implementation generates a `flow_id` internally first before writing it to an appropriate field in packet headers similar to Option-1.
+
+Generating the `flow_id` internally is a matter of implementing a simple hashing functions parameterized by existing unique identifiers for publishers/subscribers within a node (for example, the RTPS entity ID).
+
+A suitable interface should be created for publishers/subscribers to obtain the unique flow ID created by the RMW. We prefer to use community help to create this interface.
```

Comment:
Why user application needs to know the flow ID? debug information? I think that user application inputs QoS requirement, and the rest can be taken care by implementation.

Reply:
We assume that the application specifies required QoS to the network controller (e.g. Network Exposure Function in the 5GS). So it should know the unique flow ID. We prefer to avoid the coupling the RMW to the network controller to obtain the unique flow ID.

Reply:
okay i misunderstood at some point, now i see. thanks.